### PR TITLE
Dev/twegner/issue 18

### PR DESCRIPTION
**Pull request by trwegner**
With last commits i'm moving from draft to actual PR.  The changes are generally for removing the unnecessary component layer in the API referred to in issue #18.  Removed components were primarily wrappers for a single class so largely redundant.  I did encounter a number of challenges along the way since the changes introduced race conditions in several places which led to the use of concurrent-aware classes and/or locks.  As a result some tests would fail randomly due to expected order in collections no longer being stable between runs.  I was not able to find info in the spec indicated this was a problem so updated validation to allow varying order of results in the impacted tests.

Notes from draft PR:

@SergeyKanzhelev, @austinlparker, @lmolkova and @discostu105 

Looking for feedback on these changes to ensure the work is addressing #18 appropriately.  Currently still facing unit test failures in ViewManagerTest.cs and StatsRecorderTest.cs.  Any observations in this area in particular would be appreciated.

### CODE REVIEW

File: src/OpenTelemetry/Trace/Tracer.cs [hunk 1]
```diff
@@ -25,16 +25,29 @@ namespace OpenTelemetry.Trace
     public sealed class Tracer : TracerBase
     {
         private readonly SpanBuilderOptions spanBuilderOptions;
-        private readonly IExportComponent exportComponent;
         private readonly IBinaryFormat binaryFormat;
         private readonly ITextFormat textFormat;
 
-        public Tracer(IRandomGenerator randomGenerator, IStartEndHandler startEndHandler, ITraceConfig traceConfig, IExportComponent exportComponent)
-            : this(randomGenerator, startEndHandler, traceConfig, exportComponent, null, null)
```
Comment (lmolkova) [inline src/OpenTelemetry/Trace/Tracer.cs]: if it does not get export component, it should probably get exporter instead and use it in `RecordSpanData`?

Comment (trwegner) [reply src/OpenTelemetry/Trace/Tracer.cs]: Great point. Have updated implementation to include exporter.

File: src/OpenTelemetry/Trace/Tracing.cs [hunk 2]
```diff
@@ -23,32 +23,17 @@ namespace OpenTelemetry.Trace
...
-        /// <summary>
-        /// Gets the export component to upload spans to.
-        /// </summary>
-        public static IExportComponent ExportComponent => tracing.traceComponent.ExportComponent;
-
-        /// <summary>
-        /// Gets the tracer configuration.
-        /// </summary>
-        public static ITraceConfig TraceConfig => tracing.traceComponent.TraceConfig;
+        public static ITracer Tracer => TracerBase.NoopTracer;
```
[the remark this answers is not in the document]
Comment (trwegner) [reply src/OpenTelemetry/Trace/Tracing.cs]: Another great point. Have updated implementation to create a tracer.

File: test/OpenTelemetry.Tests/Impl/Trace/TracingTest.cs [hunk 3]
```diff
@@ -62,17 +62,17 @@ public void DefaultTracer()
...
 
-        [Fact(Skip = "need to fix the way tracer being instantiated")]
-        public void DefaultTraceExporter()
-        {
-            Assert.Equal(ExportComponent.NewNoopExportComponent.GetType(), Tracing.ExportComponent.GetType());
-        }
+        //[Fact(Skip = "need to fix the way tracer being instantiated")]
+        //public void DefaultTraceExporter()
+        //{
+        //    Assert.Equal(ExportComponent.NewNoopExportComponent.GetType(), Tracing.ExportComponent.GetType());
```
Comment (lmolkova) [inline test/OpenTelemetry.Tests/Impl/Trace/TracingTest.cs]: is it intentionally commented? should it be returned back or fixed?

File: test/TestApp.AspNetCore.2.0/Startup.cs [hunk 4]
```diff
@@ -51,7 +50,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddSingleton<RequestsCollector>();
             services.AddSingleton<DependenciesCollectorOptions>(new DependenciesCollectorOptions());
             services.AddSingleton<DependenciesCollector>();
-            services.AddSingleton<IExportComponent>(Tracing.ExportComponent);
+            services.AddSingleton<IExportComponent>(ExportComponent.NewNoopExportComponent);
```
Comment (lmolkova) [inline test/TestApp.AspNetCore.2.0/Startup.cs]: i probably miss something - why do we need to register noop? It doesn't seem right

File: src/OpenTelemetry/Trace/Tracing.cs [hunk 5]
```diff
@@ -23,32 +23,59 @@ namespace OpenTelemetry.Trace
...
-            this.traceComponent = new TraceComponent(new RandomGenerator(), new SimpleEventQueue());
+            IRandomGenerator randomHandler = new RandomGenerator();
+            IEventQueue eventQueue = new SimpleEventQueue();
+
+            this.TraceConfig = new Config.TraceConfig();
+
+            // TODO(bdrutu): Add a config/argument for supportInProcessStores.
+            if (eventQueue is SimpleEventQueue)
+            {
+                this.ExportComponent = Export.ExportComponent.CreateWithoutInProcessStores(eventQueue);
```
Comment (lmolkova) [inline src/OpenTelemetry/Trace/Tracing.cs]: so the only reason for export component are span stores?
But they are not even defined in the spec. 


May be @SergeyKanzhelev  could help here - why do we keep them in this repo?

File: test/OpenTelemetry.Tests/Impl/Trace/TracingTest.cs (hunk 3 above)
Comment (trwegner) [reply test/OpenTelemetry.Tests/Impl/Trace/TracingTest.cs]: Updated to restore tests

File: test/TestApp.AspNetCore.2.0/Startup.cs (hunk 4 above)
Comment (trwegner) [reply test/TestApp.AspNetCore.2.0/Startup.cs]: Reverted to original

File: src/OpenTelemetry/Trace/Tracing.cs (hunk 5 above)
Comment (trwegner) [reply src/OpenTelemetry/Trace/Tracing.cs]: Are there additional changes I should make to address ExportComponent?  It was not called out in the original issue.  Should it be another issue instead?

Comment (lmolkova) [reply src/OpenTelemetry/Trace/Tracing.cs]: sure! I didn't realize it was not par of original issue, sorry about that

Comment (trwegner) [reply src/OpenTelemetry/Trace/Tracing.cs]: No problem :)  I will add a new issue for removing ExportComponent.

**Pull request merged**